Disable DB related tests, allow to give more then 1 subtask for task, fix task chain executor tests and add copy constructor to JPetStatistics

kamilrakoczy opened the pull request:
I commented all tests in JPetDBParamGetterTest.

JPetScopeLoader now only allows to have 1 subtask and owner of JPetStatistics of subtask is JPetScopeLoader.

I added to JPetStatistics copy constructor to allow creating copy of it for subtasks in JPetTaskIO.

In JPetTaskChainExecutor I fixed error condition and also fixed JPetTaskChainExecutorTests.

In JPetTaskIO I changed saving of JPetStatistics to save statistics of all subtasks.

Commit:
Comment DB related tests

Commit:
Change JPetTaskInterface to allow give more then 1 subtask for task

wkrzemien reviewed (commented):
I don't understand why do you want to substitute unique::ptr with shared::ptr ? The original idea was that since a task which has a subtask is the owner of this subtask, than it has unique_ptr to it. Now we said to extend it to many subtask, but still the ownership relation didn't change and unique ptrs should be kept. Or I am missing something?

wkrzemien commented on JPetTaskInterface/JPetTaskInterface.h:
```diff
@@ -33,8 +35,8 @@ class JPetTaskInterface
-  virtual JPetTaskInterface* getSubTask() const = 0;
-  virtual std::string getName() const =0;
+  virtual void addSubTask(std::shared_ptr<JPetTaskInterface> subTask) = 0;
+  virtual std::vector<std::shared_ptr<JPetTaskInterface>> getSubTasks() const = 0;
```
Here * was used to express the fact that the caller does not take the ownership of the task.
In your version the caller of getSubTasks becomes a co-owner of tasks.

wkrzemien commented on JPetTaskIO/JPetTaskIO.cpp:
```diff
@@ -207,10 +211,12 @@ bool JPetTaskIO::createOutputObjects(const char* outputFilename)
-    task->setStatistics(std::move(fStatistics));
+    for (auto fSubTask : fSubTasks) {
+      auto task = dynamic_cast<JPetUserTask*>(fSubTask.get());
+      task->setStatistics(std::move(fStatistics));
```
This is probably a bug, since if you have more than one task, then the first one would take fStatistics via std::move and the second one would get the nullptr.

Commit:
Change back shared_ptr to unique_ptr in subtasks

Fix returing subtasks as pointer, change fStatistics to shared_ptr

Commit:
Marge options_new of wkrzemien/options_new

kamilrakoczy commented:
I changed back shared_ptr to unique_ptr in subtasks, also I changed now fStatistics to shared_ptr to resolve std::move bug, is it ok? 

In getSubTasks() I added const to return value.

wkrzemien commented:
@alekgajos ->  can you check the logic with JPetStatistics. I am not sure what is the right approach, Is this object should be shared between every subtask, or it should get a distinct copy of it?

wkrzemien commented on JPetTask/JPetTask.cpp:
```diff
@@ -32,12 +32,12 @@ std::string JPetTask::getName() const
 }
 
-JPetTaskInterface* JPetTask::getSubTask() const
+const std::vector<std::unique_ptr<JPetTaskInterface>>* JPetTask::getSubTasks() const
```
why not to return a vector of raw pointers ? If you return a vector of unique_ptr you are still telling that the task is not an owner of subtasks anymore, it transfers all the ownership to the one that calls it.
The best would be to add a test with a one task class that has at least two subclasses  and check how it behaves.

kamilrakoczy replied:
I'm returning pointer to vector of unique_ptr, not vector itself, but if vector of raw pointers is better I can change it. I don't know what will happen when somebody will cast pointer to reference, so I will change it to vector of raw pointers

Commit:
Change fStatistics to shared_ptr

Commit:
Change getSubtasks to return vector of raw pointers

Commit:
Update TaskChainExecutorTest to new subtask version

alekgajos commented:
Every subtask must have a distinct JPetStatistics object. In the old version with only one subtask, at the level of JPetTaskIO::terminate(), the statistics object was written to a file using JPetReader::writeCollection(). The second parameter is the name of a subdirectory in a ROOT file where the contents of the statistics object will be stored. 

The point is that for every task, we want the statistics-related objects to be stored separately. That's why now in the version with multiple subtasks, every subtask needs its own JPetStatistics object. At the end of processing the tasks, contents of each JPetStatistics should be written to the output file but to a separate subdirectory with a name corresponding to the subtask which produced it.

Probably you could e.g. keep a dictionary in JPetTaskIO which would map the subtask name to a JPetStatistics object created for and passed to a certain task, and in JPetTaskIO::terminate() write everything at once with something like this:

```
for(auto & stats : statsTaskDict)
    fWriter->writeCollection(stats.second->getStatsTable(), stats.first);
```

wkrzemien commented:
Sorry, please resolve the conflicts. I had to add some fixes today.
Also, pls update the main description,
@alekgajos pls have a look at the current version with respect to JPetStatistics. If you think it is ok, i will merge it to my branch

kamilrakoczy commented:
@wkrzemien @alekgajos I didn't pushed updated JPetStatistics version still, I agree that every task need it's own JPetStatistics object, but I think I don't need to keep dict with pointers to childs task's JPetStatistics.
When each task will end terminate will be executed and then statistics will be written to file. Also child task can not even have JPetStatistics.

Also i found bug, (dynamic_cast<JPetUserTask*>((*fSubTask).get()))->getOutputEvents() can return null and then ->Clear(); will throw segmentation fault.

Next thing: JPetTask is defining subtask as JPetTaskInterface but in JPetTaskIO there is cast to JPetUserTask, is it correct? What if someone will add subtask that inheritance from JPetTaskInterface but not JPetUserTask?

Commit:
Marge wkrzemien/options_new

Commit:
Add copy constructor to JPetStatistics, change JPetStatistics in JPetTaskIO

JPetStatistics of JPetTaskIO is copied for each subtask and is owned by JPetTaskIO

Commit:
Add method to setOutputEvents, setOutputEvents in run

Commit:
Fix writing main stats to file, change subtasks stats name

Some refactory and fixes to handle JPetTaskIO with no subtask

Commit:
Fix error condition in run and terminate

Commit:
Fix tests

Commit:
Fix output event assigment in run

Commit:
Change JPetStatistics to unique_ptr

kamilrakoczy commented:
@wkrzemien in JPetTaskChainExecutorTest I used unitTestData/JPetTaskChainExecutorTest/dabc_17025151847.unk.evt.root file from http://sphinx.if.uj.edu.pl/~alek/time_window_data_T/ when PR will be accepted it need to be added to sphinx sever to auto download or changed to another file.

Commit:
Marge wkrzemien/options_new

Commit:
Fix test TaskIO with no subtasks

Commit:
Fix JPetStatistics, allow only one subtask

Commit:
Fix typo

Commit:
Fix typo

Commit:
Update ClassDef number

alekgajos commented on JPetTaskIO/JPetTaskIO.cpp:
```diff
@@ -211,12 +226,22 @@ bool JPetTaskIO::createOutputObjects(const char* outputFilename)
+                          + std::to_string(i)
+                          + std::string(" stats")]
+        = std::move(std::unique_ptr<JPetStatistics>(new JPetStatistics(*fStatistics)));
+
```
I have pulled your changes and tried to run one of the examples, but it seems that the JPetStatistics pointer passed to the subtask here is null. If I add assert(fSubTasksStatistics[task->getName()].get()) before task->SetStatistics(...), the assertion fails.

Shouldn't the key which you use here be the same as above, i.e. not only task->getName() but with everything you append to it above?

alekgajos replied:
This comment mostly applies to the line right below the marked ones. Sorry for the confusion.

kamilrakoczy replied:
Yes, I fixed that, I added to subtask name information about number of subtask to distinguish stats from different subtasks, but with same name, and forgot to change that in setStatistics

Commit:
Fix subtask statistics name in setStatistics for subtask
